pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -166,26 +168,28 @@ func makeHookPod(hook *deployapi.LifecycleHook, deployment *kapi.ReplicationCont
 // HookExecutorPodClient abstracts access to pods.
 type HookExecutorPodClient interface {
 	CreatePod(namespace string, pod *kapi.Pod) (*kapi.Pod, error)
-	PodWatch(namespace, name, resourceVersion string) func() *kapi.Pod
+	PodWatch(namespace, name, resourceVersion string) (func() *kapi.Pod, chan struct{})
 }
 
 // HookExecutorPodClientImpl is a pluggable HookExecutorPodClient.
 type HookExecutorPodClientImpl struct {
 	CreatePodFunc func(namespace string, pod *kapi.Pod) (*kapi.Pod, error)
-	PodWatchFunc  func(namespace, name, resourceVersion string) func() *kapi.Pod
+	PodWatchFunc  func(namespace, name, resourceVersion string) (func() *kapi.Pod, chan struct{})
 }
 
 func (i *HookExecutorPodClientImpl) CreatePod(namespace string, pod *kapi.Pod) (*kapi.Pod, error) {
 	return i.CreatePodFunc(namespace, pod)
 }
 
-func (i *HookExecutorPodClientImpl) PodWatch(namespace, name, resourceVersion string) func() *kapi.Pod {
+func (i *HookExecutorPodClientImpl) PodWatch(namespace, name, resourceVersion string) (func() *kapi.Pod, chan struct{}) {
 	return i.PodWatchFunc(namespace, name, resourceVersion)
 }
 
 // NewPodWatch creates a pod watching function which is backed by a
 // FIFO/reflector pair. This avoids managing watches directly.
-func NewPodWatch(client kclient.Interface, namespace, name, resourceVersion string) func() *kapi.Pod {
+// A stop channel to close the watch's reflector is also returned.
+// It is the caller's responsibility to defer closing the stop channel to prevent leaking resources.
+func NewPodWatch(client kclient.Interface, namespace, name, resourceVersion string) (func() *kapi.Pod, chan struct{}) {
```

Comment:
`HookExecutorPodClient.PodWatch` has to change one way or the other (to take or return a channel). The [consuming code](https://github.com/openshift/origin/blob/master/pkg/deploy/strategy/support/lifecycle.go#L82) doesn't call `NewPodWatch` directly.